pkg/querier/querier.go

```diff
@@ -475,3 +499,41 @@ func UseBeforeTimestampQueryable(queryable storage.Queryable, ts time.Time) Quer
 		ts:        t,
 	}
 }
+
+func validateQueryTimeRange(ctx context.Context, userID string, startMs, endMs int64, limits *validation.Overrides, maxQueryIntoFuture time.Duration) (int64, int64, error) {
+	now := model.Now()
+	startTime := model.Time(startMs)
+	endTime := model.Time(endMs)
+
+	// Clamp time range based on max query into future.
+	if maxQueryIntoFuture > 0 && endTime.After(now.Add(maxQueryIntoFuture)) {
+		origEndTime := endTime
+		endTime = now.Add(maxQueryIntoFuture)
+
+		// Make sure to log it in traces to ease debugging.
+		level.Debug(spanlogger.FromContext(ctx)).Log(
+			"msg", "the start time of the query has been manipulated because of the 'max query into future' setting",
+			"original", origEndTime, "updated", endTime)
+
+		if endTime.Before(startTime) {
+			return 0, 0, errEmptyTimeRange
+		}
```

Comment:
This check is repeated in both `if` conditions. Should we do it only once, after all manipulations?

Reply:
I think it's more future proof doing it in both places. Doesn't look an overcomplication.

Reply:
> I think it's more future proof doing it in both places

I don't quite understand what you mean by "future proof" here. (It's not overcomplication, but looks like unnecessary duplication. Plus it would cover case when input start time is already after end time).

Reply:
As of today, we can remove the `if` and move it to the end because the two validations are done one on start time and the other one on end time. If in the future we add more validation (either on start or end), we may end up in bugs which we don't realise because of an invalid time range (start > end) introduced by a previous validation.

Reply:
I see, although I think this check is pretty universal, and hopefully unit tests will warn us about newly introduced bugs. Thanks for explaining your thinking.